sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,309 @@
+
+Importantly, our smart contracting language does not allow the
+creation of variable length lists: there are no `list` or
+`cons` constructors, and buffer lengths must be statically
+defined. Under such requirements (and given that recursion is
+illegal), determining the output lengths of functions is rather
+directly achievable. To see this, we'll examine trying to compute the
+output lengths for the only functions allowed to iterate in the
+language:
+
+```
+outputLen(map f list<t>)     := Len(list<t>) * outputLen(f t)
+outputLen(filter f list<t>)  := Len(list<t>)
+outputLen(fold f list<t> s)  := Len(s)
+```
+
+Many functions within the language will output values larger than the
+function's input, _however_, these outputs will be bound by
+statically inferable constants. For example, the data function
+_fetch-entry_ will always return an object whose size is equal
+to the specified value type of the map.
+
+A complete proof for the static runtime analysis of smart contracts
+will be included with the implementation of the language.
+
+# Deploying the Smart Contract
+
+Smart contracts on the Stacks blockchain will be deployed directly as
+source code. The goal of the smart contracting language is that the
+code of the contract defines the _ground truth_ about the intended
+functionality of the contract. While seemingly banal, many systems
+chose instead to use a a compiler to translate from a a friendly
+high-level language to a lower-level language deployed on the
+blockchain. Such an architecture is needlessly dangerous. A bug in
+such a compiler could lead to a bug in a deployed smart contract when
+no such bug exists in the original source. This is problematic for
+recovery --- a hard fork to "undo" any should-have-been invalid
+transactions would be contentious and potentially create a rift in the
+community, especially as it will not be easy to deduce which contracts
+exactly were affected and for how long. In contrast, bugs in the VM
+itself present a more clear case for a hard fork: the smart contract
+was defined correctly, as everyone can see directly on the chain, but
+illegal transactions were incorrectly marked as valid.
+
+# Example: Simple Naming System
+
+To demonstrate the expressiveness of this smart contracting language,
+let's look at an example smart contract which implements a simple
+naming system with just two kinds of transactions: _preorder_
+and _register_. The requirements of the system are as follows:
+
+1. Names may only be owned by one principal
+2. A register is only allowed if there is a corresponding preorder
+   with a matching hash
+3. A register transaction must be signed by the same principal who
+   paid for the preorder
+4. A preorder must have paid at least the price of the name. Names
+   are represented as integers, and any name less than 100000 costs
+   1000 microstacks, while all other names cost 100 microstacks.
+
+In this simple scheme, names are represented by integers, but in
```

Comment:
It seems like having built-in primitives for making buffers from strings would help with the developer experience.

Reply:
Yep -- I think there's a couple places where we could add some "syntactic sugar" to make the experience a bit better, so that buffer literals in the language could be written as strings. You'd also want library support for easily making buffers from strings. For example, in a library for constructing and signing a transaction, the library would convert the string into a buffer for sending to the network. For reading contract state, we'd want to provide a similar functionality.

Reply:
Yes, it makes sense to have most of that sugar be in the libraries for signing and reading contracts.

When defining the contract, if you were to use buffers to 'represent' strings (as here), would you use a buffer of fixed length, essentially capping the max length of the string? Is there a max buffer size in the spec?

Reply:
Added notes for the max buffer size in the cost section of the document, and added a note about syntactic improvements and library support as well -- basically by saying we should leave the door open to syntactic improvements outside of this SIP.